### ISSUE TITLE
Blueprint should not assume certain things about the environment

### ISSUE THREAD

**Issue by itsvic-dev** [BlueprintFramework/main#29]
Tested with `alpha-SLX`:
- Blueprint assumes that `/usr/local/bin` exists and is user-writable. 
- Blueprint assumes that the panel lives in `/var/www/pterodactyl`.
- Blueprint assumes that the panel directory will never change.

This causes the following issues:
- Certain distros do not have a `/usr/local` directory (such as NixOS), or even a `/usr/bin`! (NixOS has it, but it only contains `env`) Because of this, the `blueprint` tool does not get installed if `/usr/local/bin` is not present or if it's not writable to by users.
- The installer falls apart anywhere outside of `/var/www/pterodactyl`. A safer assumption would be setting `FOLDER="$(dirname $0)"`.
- When setting up symlinks to `.blueprint`, Blueprint will create absolute links. If the panel directory is renamed, Blueprint will partially break.

**Comment by prplwtf** [comment]
In all honestly, I doubt people would run Pterodactyl on NixOS.

**Comment by itsvic-dev** [comment]
> In all honestly, I doubt people would run Pterodactyl on NixOS.

I do -- but that's not the point. Point is, Blueprint shouldn't assume certain things about the environment. Another one is, Blueprint assumes that www-data is the owner of the panel directory. Doesn't break anything, but is annoying nonetheless.

**Comment by Ahwxorg** [comment]
Works on my machine

**Comment by prplwtf** [comment]
Non-standard directory support has been implemented since ages ago.

Non-standard file owner has been implemented recently, but is waiting for the next release.

**Comment by itsvic-dev** [comment]
> Non-standard directory support has been implemented since ages ago.

It's a bit wonky in my experience. When running with `_FOLDER` set, it breaks. Running it a second time, with or without `_FOLDER`, gets it to actually install.
This still doesn't address my initial complaint that the directory can change, and you should avoid absolute symlinks.

> Non-standard file ownership has been implemented recently, but is waiting for the next release.

Glad to hear that.

> we/you could just add a simple if-statement, or tell people that NixOS is unsupported?

This isn't a NixOS-specific issue. If the user running Blueprint does not have write permissions to `/usr/local/bin`, it will also fail to install the `blueprint` tool, as was the case for me as I was running it as myself, not as root.

**Comment by prplwtf** [comment]
We could switch to a .env-like config file for Blueprint which would store settings like these in a readable manner:

For the "blueprint" command issue, we could just add a statement to check from a list of directories, and if it exists, place it in there.

**Comment by prplwtf** [comment]
`_FOLDER` functionality has now been removed in commit [`6faa0c3`](https://github.com/BlueprintFramework/framework/commit/6faa0c365b98779c05d2f97e1deab7c77f90323d) in favor of :sparkles: sed commands :sparkles:.

**Comment by 0x7d8** [comment]
@prplwtf couldnt this be solved by making a shell alias for blueprint on environments that dont have the directories?

**Comment by prplwtf** [comment]
@itsvic-dev 

Checking in; are there any changes you'd like to see implemented in the next version?

**Comment by itsvic-dev** [comment]
> @itsvic-dev 
> 
> Checking in; are there any changes you'd like to see implemented in the next version?

not sure
